### DISCUSSION TITLE
kotlinx-serialization support

### DISCUSSION THREAD

**Discussion by dariuszkuc** [tschuchortdev/kotlin-compile-testing#241] [Q&A] [answered]
Hello 👋 
Was just wondering whether it would be possible to add some documentation on how to configure [`kotllinx-serialization`](https://github.com/Kotlin/kotlinx.serialization) compiler plugin? Unsure if it is supported as I haven't found any references to `ComponentRegistrar` in `kotlinx` repo.

```kotlin
val compilationResult = KotlinCompilation().apply {
  jvmTarget = "1.8"
  sources = myGeneratedSources
  inheritClassPath = true
  compilerPlugins = ???
}.compile()
```

Thanks for the great project!

**Comment by tschuchortdev** [comment, marked as answer]
I'm not sure if it is possible or how to do it. The best way to find out would be to find the code of the corresponding Gradle plugin and see what it does. I couldn't find any such code in the repository that you linked, so I suspect that the code for the Gradle plugin is somewhere else. Perhaps [this](https://github.com/JetBrains/kotlin/blob/master/plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/extensions/SerializationComponentRegistrar.kt) and [this](https://github.com/JetBrains/kotlin/blob/master/libraries/tools/kotlin-serialization/src/main/kotlin/org/jetbrains/kotlinx/serialization/gradle/SerializationSubplugin.kt) is what you're looking for? Hopefully, you just need to add the `ComponentRegistrar` that I linked to the `KotlinCompilation` and it should work, but the `ComponentRegistrar` won't be included in the artifact of the runtime library. I'm not sure if it even gets published, but you could try [this artifact](https://plugins.gradle.org/plugin/org.jetbrains.kotlin.plugin.serialization).

If you find out how to do it, you're welcome to make a PR to add documentation, or document it in this issue. I'm sure other people will find it helpful as well.

**Reply by dariuszkuc** [reply]
Indeed since it is a compiler plugin it is a part of the main kotlin repo. Based https://youtrack.jetbrains.com/issue/KT-47900 I tried using `kotlin-serialization-unshaded` (which contains kotlinx `SerializationComponentRegistrar`) but I am hitting a following problem

```
e: java.lang.IllegalStateException: The provided plugin com.tschuchort.compiletesting.MainComponentRegistrar is not compatible with this version of compiler
...
Caused by: java.lang.AbstractMethodError: Receiver class org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationComponentRegistrar does not define or inherit an implementation of the resolved method 'abstract void registerProjectComponents(org.jetbrains.kotlin.com.intellij.mock.MockProject, org.jetbrains.kotlin.config.CompilerConfiguration)' of interface org.jetbrains.kotlin.compiler.plugin.ComponentRegistrar.
````

Tried it with `com.github.tschuchortdev:kotlin-compile-testing:1.4.5` (which depends on embeddable compiler v1.5.31) and `org.jetbrains.kotlin:kotlin-serialization-unshaded:1.5.31`. Looking at the sources I see that `SerializationComponentRegistrar` does implement the required method.

```kotlin
override fun registerProjectComponents(project: MockProject, configuration: CompilerConfiguration) {
   registerExtensions(project)
}
```

 I'll try with latest Kotlin 1.6 later.

**Reply by dariuszkuc** [reply]
Ah its an issue with shaded jar.... `com.intellij.mock.MockProject` (shaded) vs `org.jetbrains.kotlin.com.intellij.mock.MockProject`

**Reply by dariuszkuc** [reply]
Tried using gradle serialization plugin as a dependency but sadly it also fails with the same error.

**Reply by tschuchortdev** [reply]
Have you tried this?
```
testImplementation "org.jetbrains.kotlin:kotlin-serialization:1.6.10"
testImplementation "org.jetbrains.kotlinx:kotlinx-serialization-json:1.3.2"
```
Seems to work for me, at first glance.

**Reply by dariuszkuc** [reply]
Looks like my issues were from some Gradle build cache as it started working today.

```kotlin
val compilationResult = KotlinCompilation().apply {
  jvmTarget = "1.8"
  sources = myGeneratedSources
  inheritClassPath = true
  compilerPlugins = listOf(SerializationComponentRegistrar())
}.compile()
```

Dependencies 

```
testImplementation("com.github.tschuchortdev:kotlin-compile-testing:1.4.5")
testImplementation("org.jetbrains.kotlin:kotlin-serialization:1.5.31")
```

Thanks for the help and awesome library.

**Reply by nsragow** [reply]
Do you have a working project? I am running into the same 'com.intellij.mock.MockProject (shaded) vs org.jetbrains.kotlin.com.intellij.mock.MockProject' issue

**Reply by dariuszkuc** [reply]
@nsragow double check on your project dependencies (and clear caches as needed) and verify you are using libraries with compatible kotlin version. ([link to my test class](https://github.com/ExpediaGroup/graphql-kotlin/blob/master/plugins/client/graphql-kotlin-client-generator/src/test/kotlin/com/expediagroup/graphql/plugin/client/generator/GraphQLTestUtils.kt#L76))
